pkg/providers/amifamily/ami.go

```diff
@@ -126,163 +82,131 @@ func (p *DefaultProvider) List(ctx context.Context, nodeClass *v1beta1.EC2NodeCl
 				// Removing this default would cause Karpenter to discover publicly shared AMIs passing the name filter.
-				elem = FiltersAndOwners{
+				query = AMIQuery{
 					Owners: lo.Ternary(term.Owner != "", []string{term.Owner}, []string{"self", "amazon"}),
 				}
-				elem.Filters = append(elem.Filters, &ec2.Filter{
+				query.Filters = append(query.Filters, &ec2.Filter{
 					Name:   aws.String("name"),
 					Values: aws.StringSlice([]string{term.Name}),
 				})
 
 			}
 			for k, v := range term.Tags {
 				if v == "*" {
-					elem.Filters = append(elem.Filters, &ec2.Filter{
+					query.Filters = append(query.Filters, &ec2.Filter{
 						Name:   aws.String("tag-key"),
 						Values: []*string{aws.String(k)},
 					})
 				} else {
-					elem.Filters = append(elem.Filters, &ec2.Filter{
+					query.Filters = append(query.Filters, &ec2.Filter{
 						Name:   aws.String(fmt.Sprintf("tag:%s", k)),
 						Values: []*string{aws.String(v)},
 					})
 				}
 			}
-			res = append(res, elem)
+			queries = append(queries, query)
 		}
 	}
 	if len(idFilter.Values) > 0 {
-		res = append(res, FiltersAndOwners{Filters: []*ec2.Filter{idFilter}})
+		queries = append(queries, AMIQuery{Filters: []*ec2.Filter{idFilter}})
+	}
+	return queries, nil
+}
+
+func (p *DefaultProvider) getAMIs(ctx context.Context, queries []AMIQuery) (AMIs, error) {
+	hash, err := hashstructure.Hash(queries, hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true})
+	if err != nil {
+		return nil, err
+	}
+	if images, ok := p.cache.Get(fmt.Sprintf("%d", hash)); ok {
+		// Ensure what's returned from this function is a deep-copy of AMIs so alterations
+		// to the data don't affect the original
+		return append(AMIs{}, images.(AMIs)...), nil
 	}
-	return res
+	images := map[uint64]AMI{}
+	for _, query := range queries {
+		if err = p.ec2api.DescribeImagesPagesWithContext(ctx, query.DescribeImagesInput(), func(page *ec2.DescribeImagesOutput, _ bool) bool {
+			for _, image := range page.Images {
+				arch, ok := v1beta1.AWSToKubeArchitectures[lo.FromPtr(image.Architecture)]
+				if !ok {
+					continue
+				}
+				requirementSets := func() []scheduling.Requirements {
+					if knownRequirements, ok := query.KnownRequirements[lo.FromPtr(image.ImageId)]; ok {
+						return lo.Map(knownRequirements, func(r scheduling.Requirements, _ int) scheduling.Requirements {
+							r.Add(scheduling.NewRequirement(v1.LabelArchStable, v1.NodeSelectorOpIn, arch))
+							return r
+						})
+					}
+					return []scheduling.Requirements{scheduling.NewRequirements(scheduling.NewRequirement(v1.LabelArchStable, v1.NodeSelectorOpIn, arch))}
+				}()
+				for _, reqs := range requirementSets {
+					reqsHash := lo.Must(hashstructure.Hash(reqs.NodeSelectorRequirements(), hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true}))
+					// If the proposed image is newer, store it so that we can return it
```

Comment:
This comment should really be describing the below, right?
```suggestion
					// If the proposed image has the same set of requirements, replace the older image with the newer image
```